### PR TITLE
BUG: string methods with NA

### LINKED ISSUE #31632: Bug in Series.str.repeat with string dtype and sequence of repeats
#### Code Sample, a copy-pastable example if possible

```pytb
In [1]: import pandas as pd

In [2]: s = pd.Series(['a', None], dtype="string")

In [3]: s.str.repeat([1, 2])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/sandbox/pandas/pandas/core/strings.py in rep(x, r)
    781             try:
--> 782                 return bytes.__mul__(x, r)
    783             except TypeError:

TypeError: descriptor '__mul__' requires a 'bytes' object but received a 'NAType'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-3-a01827562f7a> in <module>
----> 1 s.str.repeat([1, 2])

~/sandbox/pandas/pandas/core/strings.py in wrapper(self, *args, **kwargs)
   1950                 )
   1951                 raise TypeError(msg)
-> 1952             return func(self, *args, **kwargs)
   1953
   1954         wrapper.__name__ = func_name

~/sandbox/pandas/pandas/core/strings.py in repeat(self, repeats)
   2780     @forbid_nonstring_types(["bytes"])
   2781     def repeat(self, repeats):
-> 2782         result = str_repeat(self._parent, repeats)
   2783         return self._wrap_result(result)
   2784

~/sandbox/pandas/pandas/core/strings.py in str_repeat(arr, repeats)
    785
    786         repeats = np.asarray(repeats, dtype=object)
--> 787         result = libops.vec_binop(com.values_from_object(arr), repeats, rep)
    788         return result
    789

~/sandbox/pandas/pandas/_libs/ops.pyx in pandas._libs.ops.vec_binop()
    239                 result[i] = y
    240             else:
--> 241                 raise
    242
    243     return maybe_convert_bool(result.base)  # `.base` to access np.ndarray

~/sandbox/pandas/pandas/_libs/ops.pyx in pandas._libs.ops.vec_binop()
    232         y = right[i]
    233         try:
--> 234             result[i] = op(x, y)
    235         except TypeError:
    236             if x is None or is_nan(x):

~/sandbox/pandas/pandas/core/strings.py in rep(x, r)
    782                 return bytes.__mul__(x, r)
    783             except TypeError:
--> 784                 return str.__mul__(x, r)
    785
    786         repeats = np.asarray(repeats, dtype=object)

TypeError: descriptor '__mul__' requires a 'str' object but received a 'NAType'
```
#### Problem description

The `str_repeat` method correctly handles NA values when `repeats` is a scalar, but fails when its a sequence.

#### Expected Output

```python
0       a
1    <NA>
dtype: string
```

### PR DESCRIPTION
**Pull request by prakhar987**
- [x] closes #31632 
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by prakhar987** [comment]
Have added `_typ` to class `NAType` so it can be identified in `is_nan()` correctly. That's one way of doing it but probably can be done better. Will update the PR according to suggestions.

**Comment by jbrockmendel** [comment]
I don’t think we want is_nan to recognize NA.

**Comment by TomAugspurger** [comment]
Agreed. To fix #31632 we'll likely want to adjust how the `na_map` function is called.

**Comment by prakhar987** [comment]
thanks for the inputs...will update this PR or raise a new one

**Comment by prakhar987** [comment]
@TomAugspurger I dont think `na_map` is called in case repeat parameter is a vector (instead vec_binop is called). We can fix the bug by checking if type is `pandas_libs.missing.NAType` and if True not calling `bytes.__mul__` or `str.__mul__` and simply returning. If that is fine will update the PR

**Comment by prakhar987** [comment]
@TomAugspurger  I think `vec_binop` is called instead of `na_map` because input is a vector. Should we implement a variant of `vec_binop` which can handle NA? Or  a shorter fix might be checking for NA before calling `bytes.__mul__` and `str.__mul__`

**Comment by TomAugspurger** [comment]
Either of those sound sensible.

File: pandas/core/strings.py [hunk 1]
```diff
@@ -778,6 +778,8 @@ def scalar_rep(x):
     else:
 
         def rep(x, r):
+            if isinstance(x, libmissing.NAType):
```
Comment (jreback) [inline pandas/core/strings.py]: do we need the same change for scalar_rep? (and a test that hits it)?

File: pandas/tests/test_strings.py [hunk 2]
```diff
@@ -1153,6 +1153,12 @@ def test_repeat(self):
         assert isinstance(rs, Series)
         tm.assert_series_equal(rs, xp)
 
+        # GH: 31632
```
Comment (jreback) [inline pandas/tests/test_strings.py]: can you make a new test (call it test_repeat_with_null)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
can you add pd.NA in tests for the other string methods, this might be a more systematic issue.

File: pandas/core/strings.py (hunk 1 above)
Comment (prakhar987) [reply pandas/core/strings.py]: If sequence passed to repeat has None it is handled...have added a test for it

File: pandas/tests/test_strings.py (hunk 2 above)
Comment (prakhar987) [reply pandas/tests/test_strings.py]: Done.

**Comment by prakhar987** [comment]
@jreback i think other string methods will be fine because most of them use `_na_map` which handles `pd.NA`. This specific issue was due to `str_repeat` using `vec_binop` for vectored inputs which didn't handle `pd.NA`. Let me know if you want the tests anyway.

File: pandas/core/strings.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/strings.py]: instead of isinstance, just use `if x is NA`

Comment (prakhar987) [reply pandas/core/strings.py]: Thanks for the suggestion. Have updated.

**Comment by prakhar987** [comment]
@jreback should i put the pd.NA tests in separate function or just add to functions that are already there?

**Review by TomAugspurger — APPROVED** [review]
> can you add pd.NA in tests for the other string methods, this might be a more systematic issue.

@jreback this was the only case method using `vec_binop`. The rest are using `_na_map` which should handle NA correctly. So I think we're OK.


Thanks @prakhar987!

**Comment by TomAugspurger** [comment]
@Meeseeksdev backport to 1.0.x

**Pull request merged**